pandas/tests/tseries/frequencies/test_inference.py

```diff
@@ -0,0 +1,406 @@
+    # see gh-8772
+    tz = tz_naive_fixture
+    idx = date_range(date_pair[0], date_pair[1], freq=freq, tz=tz)
+    assert idx.inferred_freq == freq
+
+
+def test_infer_freq_tz_transition_custom():
+    index = date_range("2013-11-03", periods=5,
+                       freq="3H").tz_localize("America/Chicago")
+    assert index.inferred_freq is None
+
+
+@pytest.mark.parametrize("data,expected", [
+    # Hourly freq in a day must result in "H"
+    (["2014-07-01 09:00", "2014-07-01 10:00", "2014-07-01 11:00",
+      "2014-07-01 12:00", "2014-07-01 13:00", "2014-07-01 14:00"], "H"),
+
+    (["2014-07-01 09:00", "2014-07-01 10:00", "2014-07-01 11:00",
+      "2014-07-01 12:00", "2014-07-01 13:00", "2014-07-01 14:00",
+      "2014-07-01 15:00", "2014-07-01 16:00", "2014-07-02 09:00",
+      "2014-07-02 10:00", "2014-07-02 11:00"], "BH"),
+    (["2014-07-04 09:00", "2014-07-04 10:00", "2014-07-04 11:00",
+      "2014-07-04 12:00", "2014-07-04 13:00", "2014-07-04 14:00",
+      "2014-07-04 15:00", "2014-07-04 16:00", "2014-07-07 09:00",
+      "2014-07-07 10:00", "2014-07-07 11:00"], "BH"),
+    (["2014-07-04 09:00", "2014-07-04 10:00", "2014-07-04 11:00",
+      "2014-07-04 12:00", "2014-07-04 13:00", "2014-07-04 14:00",
+      "2014-07-04 15:00", "2014-07-04 16:00", "2014-07-07 09:00",
+      "2014-07-07 10:00", "2014-07-07 11:00", "2014-07-07 12:00",
+      "2014-07-07 13:00", "2014-07-07 14:00", "2014-07-07 15:00",
+      "2014-07-07 16:00", "2014-07-08 09:00", "2014-07-08 10:00",
+      "2014-07-08 11:00", "2014-07-08 12:00", "2014-07-08 13:00",
+      "2014-07-08 14:00", "2014-07-08 15:00", "2014-07-08 16:00"], "BH"),
+])
+def test_infer_freq_business_hour(data, expected):
+    # see gh-7905
+    idx = DatetimeIndex(data)
+    assert idx.inferred_freq == expected
+
+
+def test_not_monotonic():
+    rng = DatetimeIndex(["1/31/2000", "1/31/2001", "1/31/2002"])
+    rng = rng[::-1]
+
+    assert rng.inferred_freq == "-1A-JAN"
+
+
+def test_non_datetime_index2():
+    rng = DatetimeIndex(["1/31/2000", "1/31/2001", "1/31/2002"])
+    vals = rng.to_pydatetime()
+
+    result = frequencies.infer_freq(vals)
+    assert result == rng.inferred_freq
+
+
+@pytest.mark.parametrize("idx", [
+    tm.makeIntIndex(10), tm.makeFloatIndex(10), tm.makePeriodIndex(10)
+])
+def test_invalid_index_types(idx):
+    msg = ("(cannot infer freq from a non-convertible)|"
+           "(Check the `freq` attribute instead of using infer_freq)")
+
+    with pytest.raises(TypeError, match=msg):
+        frequencies.infer_freq(idx)
+
+
+@pytest.mark.skipif(is_platform_windows(),
+                    reason="see gh-10822: Windows issue")
+@pytest.mark.parametrize("idx", [tm.makeStringIndex(10),
+                                 tm.makeUnicodeIndex(10)])
+def test_invalid_index_types_unicode(idx):
+    # see gh-10822
+    #
+    # Odd error message on conversions to datetime for unicode.
+    msg = r"\('Unknown string format"
```

Comment:
@pganssle : This error message is bubbling up from this portion of logic from `dateutil`:

https://github.com/dateutil/dateutil/blob/18d3f04/dateutil/parser/_parser.py#L648-L652

Is that intended to raise a `tuple` as the error message?

Reply:
Er, no, thanks for catching that. I think I just missed it in the initial review. Happy to accept a PR making it a proper string, otherwise I'll fix it myself later.

I'll note that I don't usually test the actual *error message* because I don't consider it part of the public interface (and thus don't consider changes to it backwards-incompatible so long as the error *type* doesn't change). Of course, that policy *does* lead to problems like this.

Reply:
Done in https://github.com/dateutil/dateutil/pull/881